// tdl/dist/addon.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadAddon = void 0;
const path = __importStar(require("node:path"));
const node_gyp_build_1 = __importDefault(require("node-gyp-build"));
const packageDir = path.join(__dirname, '..');
function loadAddon(libraryFile, ignoreAlreadyLoaded = false) {
    const addon = (0, node_gyp_build_1.default)(packageDir);
    const success = addon.loadTdjson(libraryFile);
    if (!success && !ignoreAlreadyLoaded)
        throw new Error('tdjson is already loaded');
    return {
        tdold: {
            create: addon.tdoCreate,
            send: addon.tdoSend,
            receive: addon.tdoReceive,
            execute: addon.tdoExecute,
            destroy: addon.tdoDestroy
        },
        tdnew: {
            init: addon.tdnInit,
            ref: addon.tdnRef,
            unref: addon.tdnUnref,
            createClientId: addon.tdnCreateClientId,
            send: addon.tdnSend,
            receive: addon.tdnReceive,
            execute: addon.tdnExecute
        },
        setLogMessageCallback: addon.setLogMessageCallback
    };
}
exports.loadAddon = loadAddon;
